packages/backend/server/src/core/permission/ac/doc.ts

```diff
@@ -0,0 +1,90 @@
+import { Injectable } from '@nestjs/common';
+
+import { DocActionDenied } from '../../../base';
+import { Models } from '../../../models';
+import {
+  DocAction,
+  DocRole,
+  fixupDocRole,
+  mapDocRoleToPermissions,
+  WorkspaceRole,
+} from '../types';
+import { AccessController, getAccessController } from './checker';
+import { Resource } from './types';
+import { WorkspaceAccessController } from './workspace';
+
+@Injectable()
+export class DocAccessController extends AccessController<'doc'> {
+  protected readonly type = 'doc';
+
+  constructor(private readonly models: Models) {
+    super();
+  }
+
+  async permissions(resource: Resource<'doc'>) {
+    const role = await this.getRole(resource.payload);
+
+    return mapDocRoleToPermissions(role);
+  }
+
+  async can(resource: Resource<'doc'>, action: DocAction) {
+    const permissions = await this.permissions(resource);
+    return permissions[action] || false;
+  }
+
+  async assert(resource: Resource<'doc'>, action: DocAction) {
+    const allow = await this.can(resource, action);
+
+    if (!allow) {
+      throw new DocActionDenied({
+        docId: resource.payload.docId,
+        spaceId: resource.payload.workspaceId,
+        action,
+      });
+    }
+  }
+
+  async getRole(payload: Resource<'doc'>['payload']): Promise<DocRole | null> {
+    const workspaceController = getAccessController(
+      'ws'
+    ) as WorkspaceAccessController;
+    const workspaceRole = await workspaceController.getRole(payload);
+
+    const userRole = await this.models.docUser.get(
+      payload.workspaceId,
+      payload.docId,
+      payload.userId
+    );
+
+    let docRole = userRole?.type ?? (null as DocRole | null);
+
+    // fallback logic
+    if (!docRole) {
```

Comment:
If the role is set to `External`, it is `0` and will enter fallback logic. It should check whether it is `null`.